modules/stochastic_tools/src/utils/GaussianProcess.C

```diff
@@ -0,0 +1,371 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "GaussianProcess.h"
+#include "FEProblemBase.h"
+
+#include <petsctao.h>
+#include <petscdmda.h>
+
+#include "libmesh/petsc_vector.h"
+#include "libmesh/petsc_matrix.h"
+
+#include <cmath>
+
+#include "MooseRandom.h"
+#include "Shuffle.h"
+
+namespace StochasticTools
+{
+
+GaussianProcess::GPOptimizerOptions::GPOptimizerOptions(const bool inp_show_optimization_details,
+                                                        const unsigned int inp_num_iter,
+                                                        const unsigned int inp_batch_size,
+                                                        const Real inp_learning_rate,
+                                                        const Real inp_b1,
+                                                        const Real inp_b2,
+                                                        const Real inp_eps,
+                                                        const Real inp_lambda)
+  : show_optimization_details(inp_show_optimization_details),
+    num_iter(inp_num_iter),
+    batch_size(inp_batch_size),
+    learning_rate(inp_learning_rate),
+    b1(inp_b1),
+    b2(inp_b2),
+    eps(inp_eps),
+    lambda(inp_lambda)
+{
+}
+
+GaussianProcess::GaussianProcess() {}
+
+void
+GaussianProcess::initialize(CovarianceFunctionBase * covariance_function,
+                            const std::vector<std::string> & params_to_tune,
+                            const std::vector<Real> & min,
+                            const std::vector<Real> & max)
+{
+  linkCovarianceFunction(covariance_function);
+  generateTuningMap(params_to_tune, min, max);
+}
+
+void
+GaussianProcess::linkCovarianceFunction(CovarianceFunctionBase * covariance_function)
+{
+  _covariance_function = covariance_function;
```

Comment:
Why not just store the object instead of duplicating its data? and then access type, name, outputs, etc from the object?

Reply:
The covariance objects belong to the Covariance system, they are user-facing. They are stored in the warehouse right now. Furthermore, multiple GaussianProcesses may use the same covariance function so storing them on a utility class might not be the way to go.